controlplane/kubeadm/internal/workload_cluster_etcd.go

```diff
@@ -289,3 +289,73 @@ func (w *Workload) ForwardEtcdLeadership(ctx context.Context, machine *clusterv1
 	}
 	return nil
 }
+
+// EtcdStatus returns the current status of the etcd cluster
+func (w *Workload) EtcdStatus(ctx context.Context) (EtcdStatus, error) {
+	nodes, err := w.getControlPlaneNodes(ctx)
+	if err != nil {
+		return EtcdStatus{}, errors.Wrap(err, "failed to list control plane nodes")
+	}
+
+	etcdClient, err := w.etcdClientGenerator.forNodes(ctx, nodes.Items)
+	if err != nil {
+		return EtcdStatus{}, errors.Wrap(err, "failed to create etcd client")
+	}
+
+	members, err := etcdClient.Members(ctx)
+	if err != nil {
+		return EtcdStatus{}, errors.Wrap(err, "failed to list etcd members using etcd client")
+	}
+
+	statuses := map[string]EtcdMemberStatus{}
+	for _, member := range members {
+		status := EtcdMemberStatus{Member: member, Responsive: true}
+
+		_, err := etcdClient.EtcdClient.Status(ctx, staticPodName("etcd", member.Name))
+		if err != nil {
+			status.Responsive = false
+			w.Log.Error(err, "error fetching etcd status for member", "member", member.Name)
+		}
+		statuses[member.Name] = status
+	}
+
+	return EtcdStatus{
+		Members: statuses,
+	}, nil
+}
+
+// EtcdStatus is a snapshot of the etcd cluster's status
+type EtcdStatus struct {
+	Members map[string]EtcdMemberStatus
+}
+
+// EtcdMemberStatus is a snapshot of a etcd member's status
+type EtcdMemberStatus struct {
+	*etcd.Member
+	Responsive bool
+}
+
+// FailureTolerance is the amount of members the etcd cluster can afford to
+// lose without losing quorum.
+// Ref: https://github.com/etcd-io/etcd/blob/master/Documentation/faq.md#what-is-failure-tolerance
+func (e *EtcdStatus) FailureTolerance(machine *clusterv1.Machine) int {
+	var unresponsive int
+	for _, m := range e.Members {
+		if !m.Responsive {
+			unresponsive++
+		}
+	}
+
+	defaultTolerance := len(e.Members) - (len(e.Members)/2.0 + 1) - unresponsive
+
+	if machine.Status.NodeRef == nil {
+		return defaultTolerance
+	}
```

Comment:
I believe this is safe based on the ordering of operations for kubeadm join, but it may be good to add a comment to that effect, since technically a static pod deployment could be running even if there is no Node in the cluster yet.

Reply:
I decided to do this just because we can't determine the responsiveness of it if it doesn't have a NodeRef. As far as I can imagine it wouldn't make sense that a machine without a NodeRef could be running the etcd static pod and therefore be an etcd member, but I'm not certain of that.

I don't entirely know what comment to add to represent that 😸